x-pack/plugins/security_solution/docs/prebuilt_rules_customization_rfc.md

```diff
@@ -0,0 +1,728 @@
+
+### Migration strategy: perform migration on all write operations of the rules
+
+In order to migrate our users rule's data to the new schema, we should perform the migration of rules whenever any endpoint operation that writes/updates a rule saved object.
+
+> In Elasticsearch, the type of rule saved objects is called `alert`.
+
+This migration strategy of write/update operations means that a user's data will be migrated incrementally, and that both types of data (non-migrated and migrated saved objects) will coexist for an indeterminate amount of time. Therefore we have to maintain backwards data compatibility for the non-migrated data types.
+
+The migration logic should take place as a first step within the handler logic of all endpoints that carry out write/update endpoint operations, and the endpoint should return the already-migrated rule(s).
+
+### Rule Management endpoints that should include this migration logic are:
+
+All of the following endpoints either fetch the rule before updating it, or send the rule's params as part of the body in the request. Therefore, in order to carry out the migration, we can check if the `params` fields already contains the `prebuilt` and `customized` fields, create them within the `params` if they don't, and then proceed to saving them to Elasticsearch.
+
+- **Update Rule** - `PUT /rules`: same as Patch Rule.
+    - Used **only** when updating/modifying a single rule via the Rule Details page
+- **Patch Rule** - `PATCH /rules`: should migrate patched rule to prevent the endpoint from returning an outdated (non-migrated) rule.
+    - Used for attaching shared exceptions list to rules
+- **Bulk Update Rules** - `PUT /rules/_bulk_update`: same as for Patch Rule
+    - Deprecated and unused by the UI (might still be used by API users)
+- **Bulk Patch Rules** - `PATCH /rules/_bulk_update`: same as for Patch Rule
+    - Deprecated and unused by the UI (might still be used by API users)
+- **Import Rules** - `POST /rules/_import`: 
+    - if the user imports rules with the updated rule schema (`prebuilt` and `customized` fields are present), then save rule as-is. 
+    - if the user imports a "legacy" rule (with the `immutable` field, and no `prebuilt` or `customized` fields), the `immutable` field will be saved for backwards compatibility as well, and the `prebuilt` and `customized` fields should always be set to `false` (importing a non-migrated rule means that the rule could not be modified or exported in the Kibana version where it was exported from, so it could not have been a prebuilt rule).
+- **Bulk Actions** - `POST /rules/_bulk_action`: 
+    - This endpoint includes many types of actions, and only a subset of them should perform a migration.
+    - This endpoint also includes a `dry_run` mode that is executed to evaluate preconditions and warn the user before executing the actual request. No migration logic should take place for dry run requests, i.e when `dry_run=true`.
```

Comment:
> No migration logic should take place for dry run requests, i.e when `dry_run=true`.

Why the logic should be different depending on this parameter's value? Shouldn't it only affect the fact whether we save these changes or not?

Reply:
Because when the `dry_run` parameter is set to `true`, none of the the logic paths of the current endpoint code save changes to ES.

I'm adding this to the sentence, for clarity.